### ISSUE TITLE
Finer control on metrics computation

### ISSUE THREAD

**Issue by nicolaspi** [keras-team/keras#16809]
**System information**.

TensorFlow version (you are using): Nightly
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

The metrics are computed at each step during train/evaluation. While the `update_state()` calls are necessary, the calls to `result()` may not. In some cases, when the number of metrics is significant, or when the `result` method is computationally intensive, e.g. AUC: https://github.com/keras-team/keras/blob/ba2a55f1521cfebb4c9931eec50cf92530df8190/keras/metrics/metrics.py#L2028
we might want to disable the computation of metrics at the batch/step level.

Leveraging the computational cost of metrics could also be done by disabling them during training.
Cf.
https://stackoverflow.com/questions/71412499/how-to-prevent-keras-from-computing-metrics-during-training/
https://stackoverflow.com/questions/56826495/how-to-make-keras-compute-a-certain-metric-on-validation-data-only/

**Will this change the current api? How?**

`Model.compute_metrics` is augmented with the argument `training`

**Who will benefit from this feature?**

Use cases where the metrics are computationally intensive.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes
- Briefly describe your candidate solution(if contributing):

`compute_metrics` gains the boolean argument `training` indicating whether the call is made during train or evaluation.
The metrics results are gathered and put into the epoch logs, this being done independently of the behavior of `compute_metrics`, enabling the latter to be overloaded in a subclass that could then disable metric computations on the batch level or depending on the train/eval context.

This would enable a user to subclass `Metric` and `Model`, and overload `compute_metrics`, to disable metric computation on the batch level or depending on the training regime:

```python
class CustomMetric(metrics_module.BinaryAccuracy):
    def __init__(self, name, disable_train_compute):
        super().__init__(name=name)
        self.disable_train_compute = disable_train_compute

class MyModel(sequential.Sequential):
    def __init__(self, layers, disable_batch_compute):
        super().__init__(layers)
        self._disable_batch_compute = disable_batch_compute

    def compile(self, *args, **kwargs):
        metrics = kwargs.get("metrics", None)
        super().compile(*args, **kwargs)
        on_train_metrics = [
            m for m in metrics if not m.disable_train_compute
        ]
        self.on_train_compiled_metrics = compile_utils.MetricsContainer(
            on_train_metrics,
            weighted_metrics=None,
            output_names=self.output_names,
            from_serialized=False,
        )

    @property
    def loss_and_layer_metrics(self):
        metrics = []
        if self._is_compiled:
            if self.compiled_loss is not None:
                metrics += self.compiled_loss.metrics

        for l in self._flatten_layers():
            metrics.extend(l._metrics)
        return metrics

    def compute_metrics(
        self, x, y, y_pred, sample_weight, training=None
    ):
        if training:
            self.on_train_compiled_metrics.update_state(
                y, y_pred, sample_weight
            )
            metrics = (
                self.on_train_compiled_metrics.metrics
                + self.loss_and_layer_metrics
            )
            if not self.compiled_metrics.built:
                self.compiled_metrics.build(y_pred, y)
        else:
            self.compiled_metrics.update_state(y, y_pred, sample_weight)
            metrics = self.metrics

        if self._disable_batch_compute:
            return self._compute_metrics_result(
                self.loss_and_layer_metrics
            )
        else:
            return self._compute_metrics_result(metrics)
```

**Comment by nicolaspi** [comment]
Please consider this proposal as a first approach: 
https://github.com/keras-team/keras/compare/master...nicolaspi:keras:feature/metric_per_bacth_option

This would change the API by removing `Model.compute_metrics`, and remove some metric logics within `train_step` and `test_step`, they no longer return the metric values.

**Comment by divyashreepathihalli** [comment]
Closing this since this has been submitted as a PR and is currently being reviewed here - https://github.com/keras-team/keras/pull/16816
